vespalib/src/tests/ref_counted/ref_counted_test.cpp

```diff
@@ -0,0 +1,218 @@
+    CheckObjects check(1, 1);
+    auto ref = make_ref_counted<Leaf>(10);
+    Leaf &leaf = *ref;
+    Base &base = leaf;
+    EXPECT_EQ(ref->count_refs(), 1);
+    auto from_leaf = ref_counted_from(leaf);
+    static_assert(std::same_as<decltype(from_leaf),ref_counted<Leaf>>);
+    auto from_base = ref_counted_from(base);
+    static_assert(std::same_as<decltype(from_base),ref_counted<Base>>);
+    EXPECT_EQ(ref->count_refs(), 3);
+    EXPECT_EQ(from_base->val, 10);
+}
+
+TEST(RefCountedTest, use_internal_api) {
+    CheckObjects check(1);
+    Base *raw = new Base(20);
+    EXPECT_EQ(raw->count_refs(), 1);
+    ref_counted<Base> ref = ref_counted<Base>::internal_attach(raw);
+    EXPECT_EQ(ref->count_refs(), 1);
+    EXPECT_EQ(ref.internal_detach(), raw);
+    EXPECT_EQ(raw->count_refs(), 1);
+    raw->internal_addref();
+    EXPECT_EQ(raw->count_refs(), 2);
+    raw->internal_subref();
+    EXPECT_EQ(raw->count_refs(), 1);
+    raw->internal_subref();
+}
+
+TEST(RefCountedTest, move_ref_counted) {
+    for (bool has_src: {true, false}) {
+        for (bool has_dst: {true, false}) {
+            for (bool same: {true, false}) {
+                if (same) {
+                    CheckObjects check(has_src + has_dst);
+                    ref_counted<Base> src = has_src ? make_ref_counted<Base>(10) : ref_counted<Base>();
+                    ref_counted<Base> dst = has_dst ? make_ref_counted<Base>(20) : ref_counted<Base>();
+                    dst = std::move(src);
+                    EXPECT_EQ(dst, has_src);
+                    EXPECT_FALSE(src);
+                    if (has_src) {
+                        EXPECT_EQ(dst->val, 10);
+                        EXPECT_EQ(dst->count_refs(), 1);
+                    }
+                } else {
+                    CheckObjects check(has_src + has_dst, has_src + has_dst);
+                    ref_counted<Leaf> src = has_src ? make_ref_counted<Leaf>(10) : ref_counted<Leaf>();
+                    ref_counted<Base> dst = has_dst ? make_ref_counted<Leaf>(20) : ref_counted<Leaf>();
+                    dst = std::move(src);
+                    EXPECT_EQ(dst, has_src);
+                    EXPECT_FALSE(src);
+                    if (has_src) {
+                        EXPECT_EQ(dst->val, 10);
+                        EXPECT_EQ(dst->count_refs(), 1);
+                    }
+                }
+            }
+        }
+    }
+}
+
+TEST(RefCountedTest, copy_ref_counted) {
+    for (bool has_src: {true, false}) {
+        for (bool has_dst: {true, false}) {
+            for (bool same: {true, false}) {
+                if (same) {
+                    CheckObjects check(2);
+                    ref_counted<Base> empty;
+                    ref_counted<Base> obj1 = make_ref_counted<Base>(10);
+                    ref_counted<Base> obj2 = make_ref_counted<Base>(20);
+                    ref_counted<Base> src = has_src ? obj1 : empty;
+                    ref_counted<Base> dst = has_dst ? obj2 : empty;
```

Comment:
Do we want `src` and `dst` to be _refs_ to `ref_counted` instead of values to avoid any implicit reference count bumping here?

Reply:
No; calling the copy constructor with an empty/non-empty ref_counted is an intended part of the test